doc/simd_api.rst

```diff
@@ -0,0 +1,948 @@
+
+.. container:: example-code
+
+    .. code-block:: cpp
+
+        #include <util/simd.hpp>
+        using namespace arb;
+
+        void product_nonzero(int n, const double* a, const double* b, double* result) {
+            constexpr int N = simd_abi::native_width<double>::value;
+            using simd = simd<double, N>;
+            using mask = simd::simd_mask;
+
+            int i = 0;
+            for (; i+N<=n; i+=N) {
+                auto vp = simd(a+i)*simd(b+i);
+                where(vp!=0, vp).copy_to(result+i);
+            }
+
+            int tail = n-i;
+            auto m = mask::unpack((1<<tail)-1);
+
+            auto vp = simd(a+i, m)*simd(b+i, m);
+            where(m && vp!=0, vp).copy_to(c+i);
+        }
+
+
+Classes
+-------
+
+Three user-facing template classes are provided:
+
+1. ``simd<V, N, I = simd_abi::default_abi>``
+
+   *N*-wide vector type of values of type *V*, using architecture-specific
+   implementation *I*. The implementation parameter is itself a template,
+   acting as a type-map, with ``I<V, N>::type`` being the concrete implementation
+   class (see below) for *N*-wide vectors of type *V* for this architecture.
+
+   The implementation ``simd_abi::generic`` provides a ``std::array``-backed
+   implementation for arbitrary *V* and *N*, while ``simd_abi::native``
+   maps to the native architecture implementation for *V* and *N*, if
+   supported.
+
+   ``simd_abi::default_abi`` will use ``simd_abi::native`` if available, or
+   else fall back to the generic implementation.
+
+2. ``simd_mask<V, N, I = simd_api::default_abi>``
+
+   The result of performing a lane-wise comparison/test operation on
+   a ``simd<V, N, I>`` vector value. ``simd_mask`` objects support logical
+   operations and are used as arguments to ``where`` expressions.
+
+   ``simd_mask<V, N, I>`` is a type alias for ``simd<V, N, I>::simd_mask``.
+
+3. ``where_expression<simd<V, N, I>>``
+
+   The result of a ``where`` expression, used for masked assignment.
+
+Implementation typemaps live in the ``simd_abi`` namespace, while concrete
+implementation classes live in ``simd_detail``. A particular specialization
+for an architecture, for example 4-wide double on AVX, then requires:
+
+*  A concrete implementation class, e.g. ``simd_detail::avx_double4``.
+
+*  A specialization of its ABI map, so that ``simd_abi::avx<double, 4>::type``
+   is an alias for ``simd_detail::avx_double4``.
+
+*  A specialization of the native ABI map, so that
+   ``simd_abi::native<double, 4>::type`` is an alias for ``simd_abi::avx<double, 4>::type``.
+
+The maximum natively supported width for a scalar type *V* is recorded in
+``simd_abi::native_width<V>::value``.
+
+Class ``simd``
+^^^^^^^^^^^^^^
+
+The class ``simd<V, N, I>`` is an alias for ``simd_detail::simd_impl<I<V, N>::type>``;
+the class ``simd_detail::simd_impl<C>`` provides the public interface and
+arithmetic operators for a concrete implementation class `C`.
+
+In the following:
+
```

Comment:
The mixed use of italics and code quotes, i.t. *S* vs `S` is confusing. I would stick with `S` everywhere.

Reply:
I wrestled with this. Using `s` everywhere makes the text much less readable. I'm willing to try other formats for presenting this info, but would like to kick that problem down the road for a later PR.

Reply:
Fair enough, it is a hard communication challenge. The information is all there.